### ISSUE TITLE
Ability to play downloaded videos

### ISSUE THREAD

**Issue by TheDcoder** [TeamNewPipe/NewPipe#478]
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I checked if the issue/feature exists in the latest version.

Hi, I think it would be nice if NewPipe was able to play downloaded videos... much like Youtube's "Save Offline" feature.
People with phones which do not support playback of WebM might find this very useful!

@ssibb in #3932 also suggested the ability to watch the video file while it is being downloaded, in order to be able to judge the quality without wasting data by fetching the video twice (i.e. for viewing and for downloading).

**Comment by theScrabi** [comment]
Well, but this is a feature you can easily fix yourself by installing a player that can handle webm, or download the videos in mp4 format.

I'd say this is not a top level feature.

**Comment by TheDcoder** [comment]
@theScrabi Some propitiatory players might be able to do that but in my phone VLC does not work since my phone is a low-end android device 😞. I want to keep it as free as possible. Another reason for not using mp4 is that its not a free format... WebM is. and WebM files are smaller too so less data usage!

I hope you understand me :)

**Comment by turboteckel** [comment]
Isn't MPEG4 smaller than WebM in most cases on YouTube?

**Comment by TheDcoder** [comment]
@turboteckel Did not know that. When I produce my videos using Camtasia (it makes mp4) I encode it into WebM which makes the file a lot smaller without decreasing quality!

**Comment by kasbah** [comment]
Really liking NewPipe but they way this is at the moment confused me. I downloaded videos to be able to play them in background (love that feature!) but guessing from this issue when I then play them in NewPipe it is not actually playing back my downloaded videos, rather it's downloading them again by streaming them?

**Comment by kasbah** [comment]
Ah, in that case it would be good to add a "play in NewPipe" option in the downloads menu.

**Comment by theScrabi** [comment]
We need to se if that it possible with our player.

**Comment by ganguin** [comment]
Is this issue on track? It would be nice that newpipe is able to play downloaded content, especially for some newpipe features: play in the background and play in a popup. Or is there a nice an alternative audio/video that can do that?

**Comment by TheDcoder** [comment]
Still waiting for @theScrabi to confirm the possibility of this happening with the current player in NewPipe

**Comment by alayoubiyoun** [comment]
Our player of Newpipe i love it, but i want it to play just offline music with that playback(ignore silence+Tempo+Pitch) can you help plz

**Comment by opusforlife2** [comment]
@theScrabi If Newpipe stores in a database the file path of videos/audios when they are being downloaded, is it not possible for Exoplayer to play them by accessing the same file path?

**Comment by rithans** [comment]
@Stypox, 
To Use NewPipe for internal video playback , mentioned in #3083 will be implemented? Was curious as seen  #3083 marked closed.

Many has requested for the same feature, to play offline saved audio video to play it with newpipe. so where is the issue?

**Comment by TheDcoder** [comment]
@rithans That issue was closed as a duplicate. The PR (#3513) for updating ExoPlayer is still not merged, so we have to wait for it.

**Comment by voltcraft1** [comment]
sorry for more thoughts (instead of PR). I wondering if at a minimum:

Videos downloaded in NewPipe can be remembered so that a new download would be not needed when the same is played.

**Comment by desmonk** [comment]
From #4344: 

Steps to reproduce:
While watching a video, interrupt the playback with a pause. And move slider to rewind and after that the video starts to load again. Yes, that is true if we watch until the end then replay is fine. But that eventually became a offline download in cache itself.

Difference here is when a user pauses a video somewhere and goes back and forth and say does this few times then at one point the loading bar starts again kind of forgetting the cache in the process.

**Comment by Newpipeuser** [comment]
why has this feature not been  iadded yet? this was 2 years ago

**Comment by UnlimitedCookies** [comment]
This feature would be really handy since the built-in player offers double tap to rewind and a clean playback experience without having to switch apps. The photos app player really sucks in contrast. As someone already mentioned that there's a good reason the youtube app acts as the player when playing back a downloaded video.

**Comment by bew** [comment]
> The photos app player really sucks in contrast

Then don't use that _shitty_ player? Use a real video player like vlc / mxplayer / (probably others, I use mxplayer)

**Comment by opusforlife2** [comment]
> This feature would be really handy since the built-in player offers double tap to rewind

@UnlimitedCookies Give this a try: Just (Video) Player (Simple video player based on ExoPlayer library) - https://f-droid.org/packages/com.brouken.player

**Comment by UnlimitedCookies** [comment]
> > This feature would be really handy since the built-in player offers double tap to rewind
> 
> @UnlimitedCookies Give this a try: Just (Video) Player (Simple video player based on ExoPlayer library) - https://f-droid.org/packages/com.brouken.player

Just Player seems like a good video player - thanks for the recommendation!

Still it would be a better (more youtube-like) experience when the NewPipe player would be able to play downloaded videos.

**Comment by erisvulgaris** [comment]
The best feature in the new pipe is to skip silence while playing video, that is exclusively present online but if it comes available to offline videos then and student can save hours of their time while watching lecture videos please implement this offline playing feature into the new pipe.

**Comment by aminordistraction** [comment]
Although I am not much of a contributor, I personally consider an offline player as a crucial feature for NewPipe. 

1. The offline player may be used to play downloaded video AND music in-app. 
2. Music that is played could have tags, thumbnails, and attributes pulled from the app to supplement it.
3. Playlists of music may be downloaded all at once for the goal of playing them, when networks are unavailable.
4. Video, as well, may be downloaded to be played in app
5. Video, also, could have an instance of the video page downloaded for the purpose of quick archival. (save some comments, all tags, description, title, and thumbnail to preserve video info for volatile channels that have videos deleted often.)
6. Save audio/video for the purpose of enqueueing with regular video to save data
7. Automatically download select channels in the subscriptions list
8. Have the option of downloading an entire channel as an instance

It is quite disheartening to see that such a functional, and important feature has been to an extent, neglected for 3 years. Offering third party solutions for video play disregards the fact that NewPipe could offer a streamlined solution, and fulfill a function that transitioning YouTube users expect to see and use. I would offer to personally contribute to this feature, however I lack any experience in GitHub contribution, and my programming experience is very amateur. I might start, but if there are more senior developers that would be interested in this feature, I would be elated to get this into the app.

**Comment by callegar** [comment]
I think that new users of newpipe tend to get rather confused by the fact that newpipe cannot handle the videos it has saved itself.  Much of the confusion comes from the fact that newpipe has an internal file browser for these videos.

**Comment by aminordistraction** [comment]
It being a "YouTube player" would mean including in-app offline video playing in my opinion. As a previous hard-core user of the YouTube app, losing the context of a video, metadata, views, likes, some comments, really diminishes my usability of offline videos in contrast to streaming them (which half of my day is not an option). Additionally, I prefer the NewPipe player compared to any other offline player alternative I have found. My points are mainly anecdotal to how I would use this feature- I would assume other users have their own reasons for wanting this. 

As I mentioned previously, I would go to the length of coding this feature myself. However, I lack any substantial application development experience, so my contributions would likely worsen the codebase if anything. (I speak from a point of ignorance, but wouldn't playing a downloaded file in the NewPipe player basically be the same as playing a fully buffered streamed video?). I understand how ignorant I may sound on this topic, but it seems to me that much of the userbase would prefer to play offline videos in the app sometime in the future.

Not to seem combative or snarky, but for the sake of curiosity, if you are an end-user of NewPipe:
👍 - add offline playback
👎 - do not add offline playback

**Comment by opusforlife2** [comment]
> playback of videos downloaded with newpipe could be considered in scope.

@UnlimitedCookies We will need to seriously limit the user's ability to access downloads from outside of Newpipe, such as using a file manager, because even a minor, harmless looking thing like a filename change could break the hypothetical playback within Newpipe. So we could end up with something like storing videos in the private app data folder, making the user unable to access their downloads without root.

The current scenario is infinitely preferable. You download to whatever folder you want in user storage, and do whatever you want to the downloads, including playing them with your dedicated media player/music player.

**Comment by brunoais** [comment]
I believe that, for now, if I try to view a video that has already been downloaded, the app can prompt the user if the user wants to open the file instead with a video viewer.
I think that is more comfortable to do than to deal with redownloads or thinking you are downloading and can then watch later but you cannot.

In my phone, I only have space for some apps and that's it. Only 16GB doesn't give much margin of error, only 9GB, from which I need 1GB free in order to be able to update the apps

**Comment by opusforlife2** [comment]
> I believe that, for now, if I try to view a video that has already been downloaded, the app can prompt the user if the user wants to open the file instead with a video viewer.

@brunoais That is already the case. Try it.

**Comment by brunoais** [comment]
@litetex Wait what?
I thought the thumbs down was in opposition to the proposal for not doing it!

> Declare it out of scope and don't do it.

> Do we want to do this?

So, for me, the question was (below is my interpretation):
"Do we want to consider out of scope?
👍 - Agree. Out of scope for this project
😕 - no idea / neutral / abstention
👎 - Disagree, it's in scope for this project
"

Did others also vote with this philosophy?

**Comment by mhmdanas** [comment]
Probably not the team members, no.

**Comment by opusforlife2** [comment]
> Did others also vote with this philosophy?

The team was already talking about not doing it before we started posting here. So it would be quite weird if the team members suddenly took the opposite stand. xD

**Comment by brunoais** [comment]
Oh... I see... Thanks for answering

**Comment by 0Karakurt0** [comment]
Well, sad to hear it, but I do understand the reasons.
I came to this issue looking for HDR support, as of right now NewPipe's fork is the only player kapable of playing some HDR on my phone. But I don't have a good internet to stream higher resolutions...
It would  be awesome if support was added only for downloaded videos, even without file dialog or music player interface, that seems to be the lesser evil, if feature is marked as unsupported and therefore all issues are declined.

**Comment by opusforlife2** [comment]
@0Karakurt0 Try the player linked here: https://github.com/TeamNewPipe/NewPipe/issues/478#issuecomment-950059830 (https://github.com/moneytoo/Player#supported-formats says HDR is supported.)

**Comment by 0Karakurt0** [comment]
@opusforlife2 VLC also says that support it, and CxM..? Cx-something is supposedly created exactly for that but it is the only one player that managed to mangle the video beyond recognition.
But thanks, I'll try it out

**Comment by opusforlife2** [comment]
@0Karakurt0 I recommended this to you specifically because you talked about Newpipe's 'player' working, which is basically Google's Exoplayer. This app is also based on Exoplayer.

**Comment by 0Karakurt0** [comment]
@opusforlife2 thanks, works like a charm and does nothing more, awesome player.
I would love it to also support urls, but it seems to be an old issue.

**Comment by woj-tek** [comment]
Hi,
late to the party but I was just faced with similar issue. Downloading works just fine, but one of the best features / abilities of NewPipe is track what I already watched and till what point. Yes, there are external players that can play fields and even remember position but this information is lost after a while and if I go back to some video I may not remember if I watched (I use NewPipe as my all YT client with all history)

**Comment by Inwerpsel** [comment]
It's quite unfortunate to see a perfectly reasonable feature request being closed as "not planned". Just because it's not a priority at some point in time doesn't mean it's not useful or even too much work. You can save a lot of bandwidth this way if you're watching a video over and over for learning.

Since the player already has the ability to save to disk, you could could keep a database with all downloaded video IDs and a path on disk. Then, upon playback, you check if the video is in this database. Then you check if it's still on the filesystem. Finally you load the mp4 from the local file system, instead of over the network.

Maybe I'm missing something, but this doesn't seem complicated at all, and would give a lot of value.

**Comment by wizdom13** [comment]
For anyone still interested in local playback: our independent NewPipe fork, [[WizeStream](https://github.com/wizdom13/WizeStream)](https://github.com/wizdom13/WizeStream), has implemented a local audio and video browser and player.

It discovers on-device media and provides search, filters, thumbnails, background and popup playback, queue integration and details such as format, resolution, duration, capture date and audio quality.

The feature is currently implemented on WizeStream’s `pipe` branch and is planned for Android 1.7.0; it has not yet been published in a stable release. WizeStream is independent and is not affiliated with or endorsed by TeamNewPipe.
